### ISSUE TITLE
Potential ideas which could be incorporated

### ISSUE THREAD

**Issue by mraxilus** [vtomole/Keira#1]
An excerpt from the discussion on #ipfs:

> mostly just a combination of things that already exist
> 18:58
> web-based filesystem (IPFS)
> Options
> functional package management (nixOS)
> underlying service reliability (minix)
> i'm of the strong opinion that tiling widow managers (i3) are inherently better but they look like they come from the 90s
> so we really need a modern one that actually does (le gasp) any sort of smooth transitions, effects, etc.
> and looks nice out of the box
> also the representation of hardware as a filesystem, so you can use hardware over the network (plan9)

**Comment by vtomole** [comment]
Is using hardware over a network something akin to using shh to connect to a remote computer?

**Comment by mraxilus** [comment]
Sort of, but you can use any mechanism, the main idea is to access hardware through the filesystem rather than through special procedure calls. https://en.wikipedia.org/wiki/Plan_9_from_Bell_Labs provides a good explanation.

**Comment by vtomole** [comment]
Hmm, I see, so it's taking the "everything is a file" concept to the extreme. Processes will be hosted on the file system. It seems like you are leaning towards a distributed operating system which is the next obvious step after a distributed file system. I don't know how complex IPFS is, but there have been a lot of criticisms about distributed oses. https://en.wikipedia.org/wiki/Distributed_operating_system#The_price_of_complexity
What do you think?

**Comment by vtomole** [comment]
I'm sure we are not designing another distributed os though. More like a hybrid, stealing the good parts,  and leaving the bad ones.

**Comment by mraxilus** [comment]
It would make it more complex and would require someone with distributed computing experience (and security), however, it would potentially allow for some really useful ideas to exist. As an example, with AWS and Google Compute, if you need to train a machine learning model on a GPU you normally rent out time on an entire computing device, but if you only need the GPU you could get access to only that GPU. That's just one of many. But honestly, I would agree that this feature adds a lot of developmental complexity, I'm wary of the problem of creating another OS that's just slightly better than Linux in some way (like NixOS, etc.) but not a drastic enough improvement for people to actually rethink their usage of their current OS.

**Comment by vtomole** [comment]
I'm guessing the advantages to using only a part of a computer instead of the whole device saves resources/money? That would be pretty awesome. Even if it didn't, i can see how such an OS could be used by multiple people at the same time (someone uses the GPU, while another person uses the CPU). Hasn't this problem been solved already by virtualization of oses? Where multiple oses can be virtualized on the same hardware? I can kind of see how that is different from what we are trying to do.

I'm imagining something along these lines:
My GPU is represented as a hashed file on IPFS (Qmdk or something). You want to borrow it do a job of training some neural network. With something like filecoin, you wire me the currency (This can also be done after the job). 

I don't know exactly where to go from here, but i see that maybe we can manipulate the GPU file to control hardware/GPU, can this be done? If so, you can can do the job using the file, which will have a different hash after the job is done since IPFS files are immutable (it could be Qmhg), or we can make it part of IPNS and just make it mutable, maybe making it mutable would make it less secure. 

The point is everything works out without you having access to my entire system, there has to be a problem that is solved by this method i just can't think of any besides saving resources (even if that's a possibility).

**Comment by mraxilus** [comment]
The integration with filecoin is a neat idea, sort of like a peer-to-peer AWS. I'm not too familiar with the specifics as I've only had exposure to these types of technologies. It may prove to be infeasible, however, there already exists a use case for renting out your local computer for computing with things such as Folding@Home and SETI@Home. I feel like some others would also be willing to do it for filecoin. But taking a step back, it's probably best to come up with a core idea or goal to strive for, as some of these potential ideas may prove to be infeasible. Helps maintain focus.

**Comment by vtomole** [comment]
You mentioned Minix and NixOs as options. NixOs is based on the monolithic Linux, but what if the design goes along the lines of this? :https://github.com/vtomole/IPOS/blob/master/docs/brainstorming/architecture.txt

**Comment by vtomole** [comment]
Or we could just run IPFS in the kernel.

**Comment by mraxilus** [comment]
It's best to start with whatever's easiest, so we can find out what works and/or what needs to be rethought before too much time gets invested in a potential dead end.

**Comment by vtomole** [comment]
Yeah I agree. Lets start with first principles. The first step is delivering an OS that has IPFS already installed. I currently have a NixOS iso that makes it very easy to install IPFS (that's almost as close as having it installed already). All the user has to do is "cd IPOS" and then run "make". I am uploading the iso to google drive now so that i can share it with you, but it is 4GB so it will take some time. 

You currently have to run a "make" command to run the IPFS daemon. Can we make it so that the daemon is already running as soon as the OS is booted?

**Comment by vtomole** [comment]
The concept of everything being a file in UNIX is very advantageous for IPFS, because it means we can mount the whole operating system on the internet. With encryption, your OS is secure, and you can access it on any device. I see something along the lines of being able to use the same OS on your desktop, and when you leave your home, you can have access to the same system on your laptop (not remote access, but the SAME EXACT system). Call it "adapting file syncing to OS syncing". (you said "start with what's easiest", but that doesn't mean no spit-balling ideas :D)

P.S , ISO won't upload cause it's too big, but there wasn't a whole lot to show anyway, I'm currently thinking of what i should do next.

**Comment by mraxilus** [comment]
Yeah that was the idea. I built a system like that a while back using Unison/rsnyc between my laptop and PC, but it was very hacked together. It would also mean that you could use a friends computer for example and still have the same setup, etc. Really great idea.

**Comment by vtomole** [comment]
Could you go into more detail about how this hack you did worked?

**Comment by mraxilus** [comment]
I just set up Unison to sync the entirety of both computers bi-directionally. Nothing too fancy. I just hated that I had different software/files/experience on my laptop vs PC. I just wanted to have one 'computer' which I could use either device to login to it. That was early on in Linux conversion, so needless to say I messed up one system and it unintentionally synced to my other. Needless to say that messed up both systems and I lost a bunch of stuff. Haven't really bothered to do it since as I just have one fairly powerful laptop now that I use for most things, gaming included.
